File: itag_engineering/itag_engineering_management/report/forward_traceability/forward_traceability.py
# Copyright (c) 2026, Globcom Qatar and contributors
# For license information, please see license.txt

from itag_engineering.itag_engineering_management.traceability_service import forward_traceability


def execute(filters=None):
	"""A thin wrapper around traceability_service.forward_traceability() -
	the traversal logic lives entirely in that service, this report only
	flattens its "finished_units" list into rows. Requires an `identity`
	filter (a heat number, Batch, Serial No, WIP Unit Register, or
	drawing revision); with none given, returns no rows."""
	columns = [
		{
			"label": "Finished WIP Unit",
			"fieldname": "wip_unit",
			"fieldtype": "Link",
			"options": "WIP Unit Register",
			"width": 150,
		},
		{"label": "Item", "fieldname": "item", "fieldtype": "Link", "options": "Item", "width": 150},
		{
			"label": "Serial Number",
			"fieldname": "serial_number",
			"fieldtype": "Link",
			"options": "Serial No",
			"width": 150,
		},
		{"label": "Delivery Notes", "fieldname": "delivery_notes", "fieldtype": "Data", "width": 200},
		{"label": "Customer", "fieldname": "customer", "fieldtype": "Data", "width": 150},
		{
			"label": "Engineering Change Order",
			"fieldname": "eco",
			"fieldtype": "Link",
			"options": "Engineering Change Order",
			"width": 170,
		},
	]

	filters = filters or {}
	identity = filters.get("identity")
	if not identity:
		return columns, []

	result = forward_traceability(identity)
	data = []
	for unit in result.get("finished_units", []):
		data.append(
			{
				"wip_unit": unit.get("wip_unit"),
				"item": unit.get("item"),
				"serial_number": unit.get("serial_number"),
				"delivery_notes": ", ".join(unit.get("delivery_notes") or []),
				"customer": unit.get("customer"),
				"eco": unit.get("eco"),
			}
		)
	return columns, data

File: itag_engineering/itag_engineering_management/report/forward_traceability/forward_traceability.js
// Copyright (c) 2026, Globcom Qatar and contributors
// For license information, please see license.txt

// forward_traceability.py's execute() requires an `identity` filter and
// silently returns no rows without one - but no report .js existed to
// ever surface that filter field in the Desk. Same missing-UI-trigger gap
// as backward_traceability.js, fixed the same way. Plain Data field, not
// a Link, since traceability_service.forward_traceability() accepts a
// heat number, Batch, Serial No, WIP Unit Register, or drawing revision -
// five different doctypes, no single Link options value would fit.
// (execute() never forwards include_recall_population_check to the
// service call, so that parameter is intentionally left as backend-only,
// not exposed as a second filter here.)
frappe.query_reports["Forward Traceability"] = {
	filters: [
		{
			fieldname: "identity",
			label: __("Heat No / Batch / Serial No / WIP Unit / Drawing Revision"),
			fieldtype: "Data",
			reqd: 1,
		},
	],
};
